fitbenchmarking/core/fitting_benchmarking.py

```diff
@@ -65,37 +127,204 @@ def fitbenchmark_group(group_name, options, data_dir):
+    :type options: fitbenchmarking.utils.options.Options
+
+    :return: list of all results, failed problem names and dictionary of
+             unselected minimizers based on algorithm_type
+    :rtype: tuple(list of fibenchmarking.utils.fitbm_result.FittingResult,
+                  list of failed problem names,
+                  dictionary of minimizers)
+    """
+    grabbed_output = output_grabber.OutputGrabber(options)
+    results = []
+
+    software = options.software
+    if not isinstance(software, list):
+        software = [software]
+
+    problem_fails = []
+    unselected_minimzers = {}
+    software_results = []
+    for s in software:
+        LOGGER.info("        Software: %s", s.upper())
         try:
-            results_dict[name].append(problem_result)
+            minimizers = options.minimizers[s]
         except KeyError:
-            results_dict[name] = [problem_result]
+            raise UnknownMinimizerError(
+                'No minimizer given for software: {}'.format(s))
+        with grabbed_output:
+            controller_cls = ControllerFactory.create_controller(
+                software=s)
+            controller = controller_cls(problem=problem)
 
-    results = [results_dict[r] for r in
-               sorted(results_dict.keys(), key=str.lower)]
+        controller.parameter_set = start_values_index
 
-    return results, failed_problems, unselected_minimzers
+        #########################
+        # Loops over minimizers #
+        #########################
+        problem_result, minimizer_failed = \
+            loop_over_minimizers(controller=controller,
+                                 minimizers=minimizers,
+                                 options=options)
+        unselected_minimzers[s] = minimizer_failed
+        software_results.extend(problem_result)
+
+    # Checks to see if all of the minimizers raised an exception and
+    # record the problem name if that is the case
+    software_check = [np.isinf(v.chi_sq) for v in software_results]
+    if all(software_check):
+        software_results = []
+        problem_fails.append(problem.name)
+    results.extend(software_results)
+
+    return results, problem_fails, unselected_minimzers
+
+
+def loop_over_minimizers(controller, minimizers, options):
+    """
+    Loops over minimizers in fitting software
+
+    :param controller: The software controller for the fitting
+    :type controller: Object derived from BaseSoftwareController
+    :param minimizers: array of minimizers used in fitting
+    :type minimizers: list
+    :param options: FitBenchmarking options for current run
+    :type options: fitbenchmarking.utils.options.Options
+
+    :return: list of all results and dictionary of unselected minimizers
+             based on algorithm_type
+    :rtype: tuple(list of fibenchmarking.utils.fitbm_result.FittingResult,
+                  list of failed minimizers)
+    """
+    grabbed_output = output_grabber.OutputGrabber(options)
```

Comment:
I know you mentioned the other day that ideally this should be a class (which I agree with), but in the meantime should we maybe just pass the `grabbed_output` to this point (from `loop_over_benchmark_problem`) rather than having to calculate it again here?

Reply:
Agreed and corrected